### ISSUE TITLE
Clarifications around storage and training precision for 1.3B and 2.7B

### ISSUE THREAD

**Issue by nostalgebraist** [EleutherAI/gpt-neo#207]
First of all, thank you for the code and pretrained models!

I notice that the two released models differ in the precision used for weight storage:

- The 1.3B checkpoint stores model variables in `bfloat16`.
  - This is the outcome of `"precision" : "bfloat16"` in the [config](https://the-eye.eu/public/AI/gptneo-release/GPT3_XL/config.json), when interpreted by the code [here](https://github.com/EleutherAI/gpt-neo/blob/3d91883a238649f8fdb5f42e6ae9e33002bd276d/model_fns.py#L35-L39).
- The 2.7B checkpoint stores variables in `float32`.

Additionally, if the same code was used for training, then:

- The 1.3B model was trained in mixed precision, with 
  - `bfloat16` for activations, [except in the logits --> loss step](https://github.com/EleutherAI/gpt-neo/blob/67c40799d2d26abf75a5a4c00647caabdf19a42f/models/gpt2/gpt2.py#L178-L179)
  - `float32` for in-memory weights / gradients / Adam state
  - `bfloat16` for checkpoints
- The 2.7B model was trained in full precision, with `float32` everywhere.

(Sidenote: the 1.3B checkpoint has Adam state saved in `float32`, which is consistent w/ the above.)

----

I have some questions about this:

1. Why do the two models differ in this way?  I'd be less surprised if the bigger model used lower precision, but here we have the reverse.
2. _Storage_ in `bfloat16` is generally considered risky, at least for some architectures.  Do we have evidence about whether the use of `bfloat16` for checkpoints hurts the performance of the 1.3B model, relative to an identical model with `float32` storage?
3. I haven't used the [Huggingface releases](https://huggingface.co/EleutherAI/gpt-neo-1.3B), but just glancing at the configs and file sizes, I get the impression the 1.3B there is stored in 32-bit precision.
  a. Is this true?
  b. If so, was it cast to 32-bit precision from the original 16-bit checkpoint?  Or was it re-trained?
4.  I would be more likely to use these models in practice if I had a pretrained 1.3B checkpoint in 32-bit precision. (And "natively" so, not via casting from half-precision).  Does such a model exist?  Is one likely to be released later?

----

For context, right now I'm assessing the 1.3B GPT-Neo as a potential replacement for the generator in my tumblr bot (currently a 1.5B GPT-2).  It's finetuning as I type this :)

(EDIT: corrected some misuses of the terms "single" and "double" precision in my original comment.)

**Comment by nostalgebraist** [comment]
> > For context, right now I'm assessing the 1.3B GPT-Neo as a potential replacement for the generator in my tumblr bot (currently a 1.5B GPT-2). It's finetuning as I type this :)
> 
> As a sidenote, be sure to try out the 2.7B model as well. Anecdotally, people seem to like it a lot. https://twitter.com/dril_gpt2 has been powered by it since March 27th.
> 
> (I mention this because drilbot's curator tried 1.3B and almost dismissed Neo, but then was surprised by 2.7B's quality.)

I'll definitely use it if I can support the compute load, which would be the limiting factor.  (I'm a big fan of the OA scaling papers and will use a bigger GPT whenever I can)

Re: the original question, I asked it on Discord and got helpful answers:

- the bfloat16 checkpointing was an accident
- probably little to no quality impact: training loss curves don't change noticeably around load/save steps

Still, for my own purposes, I cast the 1.3B checkpoint to float32, and my further finetuning work is now checkpointing in float32.
